### PR TITLE
Return decimal for decimal-based quantities

### LINKED ISSUE #1058: Is there a reason, the conversion properties return double for quantities with base type decimal?
Some quantities, for instance `information` have `decimal` as base type (which does make sense). However, except for the `Value` property, all properties return double:

```csharp
/// <summary>
///     Gets a <see cref="double"/> value of this quantity converted into <see cref="InformationUnit.Bit"/>
/// </summary>
public double Bits => As(InformationUnit.Bit);

/// <summary>
///     Gets a <see cref="double"/> value of this quantity converted into <see cref="InformationUnit.Byte"/>
/// </summary>
public double Bytes => As(InformationUnit.Byte);
```

This is a bit confusing and unnecessarily looses precision. Would it be possible to change these properties to `decimal`, too?

### PR DESCRIPTION
**Pull request by pgrawehr**
Fixes #1058 

Return the correct value type for quantities that use `decimal` internally; `Power, Information, BitRate`.

- Quantity properties return `decimal` or `double` based on internal value type.
- `IQuantity.Value` returns `QuantityValue`, which supports both double and decimal.
- `QuantityValue`: Implement IEquality<QuantityValue>, IComparable<QuantityValue>, IComparable.

### CODE REVIEW

**Comment by angularsen** [comment]
Will get to this, just a bit short on time these days.

**Comment by pgrawehr** [comment]
@angularsen No worries, take your time. Good coders are always busy ;-)

File: UnitsNet/QuantityValue.cs [hunk 1]
```diff
@@ -90,10 +113,235 @@ private QuantityValue(decimal val)
...
+                return new QuantityValue(-v._valueDecimal.Value);
+            }
+            else
+            {
+                return new QuantityValue(-v._value.GetValueOrDefault());
+            }
+        }
+
+        /// <summary>
+        /// Multiplication operator. Note that this performs an operation on a raw value, as no unit is associate the the value at this point.
```
Comment (angularsen) [inline UnitsNet/QuantityValue.cs:290]: I initially would argue that it is redundant to explain the unit is not involved, but I realize the name `QuantityValue` is misleading. It was originally only used internally.

It might be a good idea to rename it to `Number` or something more generic in v5, to make it more clear it does not really have anything to do with quantities on its own. This would be a seaparate PR anyhow.

A minor suggestion:
```xml
<remarks>
This performs an operation on the numeric value of a quantity, there is no unit or conversions involved.
</remarks>
```

File: UnitsNet/QuantityValue.cs [hunk 2]
```diff
@@ -90,10 +113,235 @@ private QuantityValue(decimal val)
...
+        /// </summary>
+        public static QuantityValue operator *(QuantityValue a, QuantityValue b)
+        {
+            if (a._valueDecimal.HasValue && b._valueDecimal.HasValue)
+            {
+                return new QuantityValue(a._valueDecimal.Value * b._valueDecimal.Value);
+            }
+            else
```
Comment (angularsen) [inline UnitsNet/QuantityValue.cs:298]: Don't we have to check the two other `else if` cases too, to avoid calculating with unassigned values? Same below.

**Review by angularsen — CHANGES_REQUESTED** [review]
Wow, this is a ton of great work 👏 
Some suggestions below, but it's close to ready!

File: CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs [hunk 3]
```diff
@@ -882,6 +917,16 @@ public double As(UnitSystem unitSystem)
             return ToUnit(unit, DefaultConversionFunctions);
         }}
 
+        /// <summary>
+        ///     Converts this {_quantity.Name} to another {_quantity.Name} with the unit representation <paramref name=""unit"" /> and returns its see <cref name=""QuantityValue"" />.
+        /// </summary>
+        /// <param name=""unit"">The unit to convert to.</param>
+        /// <returns>A {_quantity.Name} with the specified unit.</returns>
+        public QuantityValue ToQuantity({_unitEnumName} unit)
```
[the remark this answers is not in the document]
Comment (pgrawehr) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:925]: The reason I'm trying to avoid option D is that it's error prone looking forward. 

Assume you had written (to be on the "safe" side):
```csharp
// Option D - No method, let consumer cast instead.
return MassFlow.FromKilogramsPerSecond((double)bsfc.KilogramsPerJoule * (double)power.Watts);
```
and now you decide that `BrakeSpecificFuelConsumption` (the left hand argument) should use `decimal` as base type as well. There won't be any errors or warnings, but the result of the above calculation will be inexact and nobody would notice. It gets even uglier when we'd once have more base types. 

With the options A-C, the multiplication actually uses `QuantityValue.operator*()` to perform the operation. The left hand side is implicitly converted to QuantityValue (it wouldn't hurt to do it explicitly) and we have one place to define the behavior when multiplying values of different types with each other. Otherwise, each of these multiplications or divisions would need to specify what type the result of e.g. decimal*double should be. 

About A-C: I don't have a big preference there, but we should be aware that this scenario also affects user code if he performs an operation on units that's not predefined in the library. Hence A is probably the easiest to use.

File: UnitsNet/QuantityValue.cs (hunk 1 above)
Comment (pgrawehr) [reply UnitsNet/QuantityValue.cs:290]: Ok, sounds like a plan. I added this to make it obvious that `quantityA * quantityB` is something very different from `quantityA.QuantityValue * quantityB.QuantityValue`.

File: UnitsNet/QuantityValue.cs (hunk 2 above)
Comment (pgrawehr) [reply UnitsNet/QuantityValue.cs:298]: True, I overlooked that.

File: CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs [hunk 4]
```diff
@@ -861,15 +876,35 @@ public double As(UnitSystem unitSystem)
...
+            if(unitSystem is null)
+                throw new ArgumentNullException(nameof(unitSystem));
+
+            var unitInfos = Info.GetUnitInfosFor(unitSystem.BaseUnits);
+
+            var firstUnitInfo = unitInfos.FirstOrDefault();
+            if(firstUnitInfo == null)
+                throw new ArgumentException(""No units were found for the given UnitSystem."", nameof(unitSystem));
+
+            return ((IQuantity<{_unitEnumName}>)this).As(firstUnitInfo.Value);
```
[the remark this answers is not in the document]
Comment (pgrawehr) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:896]: Done, this is the only new one IIRC.

[the remark this answers is not in the document]
Comment (angularsen) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs]: Found some others, pushed a fix.

File: CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (angularsen) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:917]: >  There won't be any errors or warnings, but the result of the above calculation will be inexact and nobody would notice. 

I agree with this sentiment, that we don't want calculations to change without anyone noticing. 

However;

1. I think it is more likely to run into changes to calculations without noticing when `QuantityValue` casts values internally. It is not intuitive for the consumer to know if they are getting double or decimal precision on their calculations.
2. You argue `we have one place to define the behavior when multiplying values of different types with each other`. This is convenient for us, but I still think it is a trap for the consumer.
3. Having to call `ToQuantityValue()` is not very intuitive. Getting a compile error on trying to multiple `double` with `decimal` is highly discoverable, although yes it requires an explicit cast.
4. It is with good reason that .NET also fails to compile in this case, instead of trying to automatically cast one way or the other.

**Comment by angularsen** [comment]
I was not able to push a change, so here is the diff I tried to push regarding As() overloads:

[0001-Simplify-As-overloads-by-reusing-and-casting-to-doub.patch.txt](https://github.com/angularsen/UnitsNet/files/8749533/0001-Simplify-As-overloads-by-reusing-and-casting-to-doub.patch.txt)

File: CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (pgrawehr) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs:916]: That sounds reasonable. 

What about we go with solution B (from the above list), which performs an explicit cast from double/decimal to QuantityValue, so the option to use `QuantityValue.operator*` is possible if desired. 

Actually, we do not need to do anything to allow explicit casts. These are always possible. Internally, we should avoid them though, to prevent the error described above.

**Comment by pgrawehr** [comment]
@angularsen There seems to be a weird problem with the use of `UnitsNetJsonConverter` that now throws a MissingMethodException I don't fully understand. Possibly due to a mix of using the interface (that has an explicit implementation of `Value` now) and the reflection that happens inside the json writer. But since that class is marked obsolete anyway, I'm not sure it's worth fixing it. Can UnitsNetJsonConverter be dumped?

**Comment by angularsen** [comment]
On the phone, but we can't make breaking changes in v4. Either we must fix
it or move this change to v5 branch.

**Comment by pgrawehr** [comment]
@angularsen The entire PR is intended to go to v5 only.

**Comment by angularsen** [comment]
Aha, my bad. 
`UnitsNetJsonConverter` seems already removed in v5 branch, but it seems the backwards compatibility tests are failing on it. I find that odd, I thought those tests explicitly solely dependend on v4 nugets in order to test backwards compatibility but maybe they do bring on v5 implementation into the mix.

I think we need to figure out why these tests are failing, because I don't want to remove the compatibility tests so we can make sure we aren't breaking anything.

https://ci.appveyor.com/project/angularsen/unitsnet/builds/43692947/tests

I don't have time to help look into it right now, but might be able to assist later this week.

**Comment by lipchev** [comment]
Hey guys, I've just had a look at this - and it appears that the only tests that are failing are from the `UnitsNet.Serialization.JsonNet.CompatibilityTests` project- which IMO should have been removed altogether in v5. It only tests the `UnitsNetJsonConverter` (the converter is already removed in v5) - via the official nuget!? Here's an extract from csproj:

    
    <PackageReference Include="UnitsNet.Serialization.JsonNet" Version="4.4.0" />

**Comment by pgrawehr** [comment]
@lipchev  Yea, it's still strange that the test fails, as it only uses old implementations (or maybe not?) I forgot about this, but I'll have a look when I find time.

**Comment by lipchev** [comment]
It's incorrect in referencing the converter via the nuget, while the UnitsNet core is referenced using a project reference. Had the reference to the converter been via direct reference it would have caused a compile time issue at the moment that the old converter was removed, thus forcing us to also remove the converter tests (i.e. the whole project)..

**Comment by pgrawehr** [comment]
@angularsen You write above that you want the compatibility tests still to pass, but given v5 doesn't even have the `UnitsNetJsonConverter`, I agree that the tests should just be removed. There's nothing we can compare compatibility with if the new implementation neither reads or writes data in this format.

**Comment by lipchev** [comment]
Although it might be argued that the tests actually did their intended purpose- if that was to test the latest UnitsNet version against UnitsNet.Serialization.JsonNet v4.4.0 - it was correct in establishing that UnitsNet v5 is not compatible with it. :)

File: CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (angularsen) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs]: Looking back at this with fresh eyes, I still lean towards solutions C or D, basically forcing the consumer to make an informed choice when mixing arithmetic of decimal vs double.

@lipchev @tmilnthorp what do you think?

```c#
// Option A - minor rename, discoverable, but must be added to all 100+ quanities
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.ToQuantityValue(PowerUnit.Watt));

// Option B - New `QuantityValue` property, less discoverable and the extra value property seems a bit weird until you know why it is useful.
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.ToUnit(PowerUnit.Watt).QuantityValue);

// Option C - Extension method on `decimal` to convert to `QuantityValue`, less discoverable, but 1 vs 100 methods on binary size
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.Watts.ToQuantityValue());

// Option D - No method, let consumer cast instead.
return MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule*(double)power.Watts);
```

**Comment by angularsen** [comment]
I just removed the compatibility tests, I agree they no longer serve a purpose.
142a1fb - 🔥 Remove UnitsNet.Serialization.JsonNet.CompatibilityTests

File: CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lipchev) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs]: I think option D is most consistent with the change of the `IQuantity.Value` type: anybody migrating to v5 would have to get used to casting on the deconstruction side: `var value = (double) quantity.Value` (or something like `quantity.Value.ToDouble()`), so I guess the same logic should apply on the construction side.

The particular line that you've highlighted however points to the other issue- around the question of what types are selected when multiplying two different types, with the options being:

- Option A, B, C: resulting in a Double * Decimal => Decimal => Double (if we follow the operator in the **public** `QuantityValue` class)
- Option D: resulting in Double * Double => Double  // note that the behavior is slightly different from what we'd get by doing `MassFlow.FromKilogramsPerSecond(bsfc.KilogramsPerJoule * power.Watts)` by hand (without casting)

On the opposite side of the operation:
A, B, C => ~ `return SpecificEnergy.FromJoulesPerKilogram(power.Watts.ToQuantityValue() / massFlow.KilogramsPerSecond);` // (Decimal / Double) => Decimal => Double
vs
D =>   `return SpecificEnergy.FromJoulesPerKilogram((double)power.Watts / massFlow.KilogramsPerSecond);` // (Double/ Double) => Double : again slightly different from power.Watts / massFlow.KilogramsPerSecond (result here is Decimal)

In short, option D is the `operation strategy™` where operands are cast to the **expected** result type prior to performing the operations, while the other ones let the resulting type vary - with a bias for Decimals (with possible overflows). 

My point is that unless we want to support **both** double & decimal as a possible `_value` for a given quantity then there is little point in relying on operators defined in the `QuantityValue` (especially since we'd be overriding all but the trivial cases).

[the remark this answers is not in the document]
Comment (pgrawehr) [reply CodeGen/Generators/UnitsNetGen/QuantityGenerator.cs]: > My point is that unless we want to support both double & decimal as a possible _value for a given quantity then there is little point in relying on operators defined in the QuantityValue (especially since we'd be overriding all but the trivial cases).

That's a fair point. I expected that multiplying a `double` with a `decimal` by first converting the double to decimal would always return the correct result. But that's not true, that multiplication can always loose accuracy or significant digits. So it's probably better to always leave the decision to the user as to how exactly to perform the conversion.

**Comment by pgrawehr** [comment]
@angularsen Hi!

I've lost track of this one a bit, but how shall we go on? It would be good if we could finally come to a conclusion (and get V5 out!)

There's also a build failure now that I think is unrelated to the PR.

**Comment by pgrawehr** [comment]
@angularsen It would be good this could be taken on hand some time soon. I'm actually waiting for a v5.0 release due to some other internal changes that make stuff easier (and of course the included fixes).

**Comment by angularsen** [comment]
Thank you for the reminder, I pushed some minor fixes and I think this is ready to merge now. Awesome job!

By the way, v5 is already out as a alpha prerelease nuget. It does lag behind master on new units though, because each merge to update that branch incur the mother of all merge conflicts. 🙈

I still have a few breaking changes I'd like to get around to before stabilizing the v5 nuget, but other than that it should be good to use already.

**Comment by angularsen** [comment]
I merged v5 into master now by the way.
Nuget on the way out.

[Release UnitsNet/5.0.0-rc001 · angularsen/UnitsNet](https://github.com/angularsen/UnitsNet/releases/tag/UnitsNet%2F5.0.0-rc001)

**Comment by pgrawehr** [comment]
> I merged v5 into master now by the way. Nuget on the way out.
> 
> [Release UnitsNet/5.0.0-rc001 · angularsen/UnitsNet](https://github.com/angularsen/UnitsNet/releases/tag/UnitsNet%2F5.0.0-rc001)

Great. I'll start testing it asap. We're preparing a major release in dotnet/iot that will bring a bunch of breaking changes, and therefore is a good time to also upgrade UnitsNet.
